File: src/test/java/burrows/apps/example/gif/presentation/main/MainPresenterTest.java
package burrows.apps.example.gif.presentation.main;

import burrows.apps.example.gif.data.rest.model.RiffsyResponse;
import burrows.apps.example.gif.data.rest.repository.RiffsyApiClient;
import burrows.apps.example.gif.presentation.IBaseSchedulerProvider;
import burrows.apps.example.gif.presentation.ImmediateSchedulerProvider;
import io.reactivex.Observable;
import org.junit.Before;
import org.junit.Test;
import org.mockito.Mock;
import test.TestBase;

import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;
import static org.mockito.MockitoAnnotations.initMocks;

/**
 * @author <a href="mailto:jaredsburrows@gmail.com">Jared Burrows</a>
 */
public final class MainPresenterTest extends TestBase {
  private final IBaseSchedulerProvider provider = new ImmediateSchedulerProvider();
  @Mock private IMainView view;
  @Mock private RiffsyApiClient repository;
  private MainPresenter sut;

  @Before @Override public void setUp() throws Throwable {
    super.setUp();

    initMocks(this);

    when(view.isActive()).thenReturn(true);
  }

  @Test public void testLoadTrendingImagesNotActive() {
    // Arrange
    final RiffsyResponse response = new RiffsyResponse();
    when(view.isActive()).thenReturn(false);
    sut = new MainPresenter(view, repository, provider);
    when(repository.getTrendingResults(any(Integer.class), any(Float.class))).thenReturn(Observable.just(response));

    // Act
    sut.loadTrendingImages(0f);

    // Assert
    verify(view).isActive();
    verify(view, times(0)).addImages(eq(response));
  }

  @Test public void testLoadTrendingImagesSuccess() {
    // Arrange
    final RiffsyResponse response = new RiffsyResponse();
    sut = new MainPresenter(view, repository, provider);
    when(repository.getTrendingResults(any(Integer.class), any(Float.class))).thenReturn(Observable.just(response));

    // Act
    sut.loadTrendingImages(0f);

    // Assert
    verify(view).isActive();
    verify(view).addImages(eq(response));
  }

  @Test public void testLoadSearchImagesSuccess() {
    // Arrange
    final RiffsyResponse response = new RiffsyResponse();
    sut = new MainPresenter(view, repository, provider);
    when(repository.getSearchResults(any(String.class), any(Integer.class), any(Float.class))).thenReturn(Observable.just(response));

    // Act
    sut.loadSearchImages("gifs", 0f);

    // Assert
    verify(view).isActive();
    verify(view).addImages(eq(response));
  }
}

File: src/test/kotlin/burrows/apps/example/gif/presentation/main/MainPresenterTest.kt
package burrows.apps.example.gif.presentation.main

import burrows.apps.example.gif.data.rest.model.RiffsyResponseDto
import burrows.apps.example.gif.data.rest.repository.RiffsyApiClient
import burrows.apps.example.gif.data.rest.repository.RiffsyApiClient.Companion.DEFAULT_LIMIT_COUNT
import com.nhaarman.mockito_kotlin.any
import com.nhaarman.mockito_kotlin.eq
import com.nhaarman.mockito_kotlin.never
import com.nhaarman.mockito_kotlin.verifyZeroInteractions
import io.reactivex.Observable
import org.junit.Before
import org.junit.Test
import org.mockito.Mock
import org.mockito.Mockito.`when`
import org.mockito.Mockito.times
import org.mockito.Mockito.verify
import org.mockito.MockitoAnnotations.initMocks
import test.ImmediateSchedulerProvider
import test.TestBase

/**
 * @author [Jared Burrows](mailto:jaredsburrows@gmail.com)
 */
class MainPresenterTest : TestBase() {
  private val provider = ImmediateSchedulerProvider()
  @Mock private lateinit var view: IMainView
  @Mock private lateinit var repository: RiffsyApiClient
  private lateinit var sut: MainPresenter

  @Before override fun setUp() {
    super.setUp()

    initMocks(this)

    `when`(view.isActive()).thenReturn(true)
  }

  @Test fun testLoadTrendingImagesNotActive() {
    // Arrange
    val next = 0
    val response = RiffsyResponseDto()
    `when`(view.isActive()).thenReturn(false)
    sut = MainPresenter(view, repository, provider)
    `when`(repository.getTrendingResults(eq(DEFAULT_LIMIT_COUNT), eq(next)))
      .thenReturn(Observable.just(response))

    // Act
    sut.loadTrendingImages(next)

    // Assert
    verify(view).isActive()
    verify(view, times(0)).addImages(eq(response))
  }

  @Test fun testLoadTrendingImagesSuccess() {
    // Arrange
    val next = 0
    val response = RiffsyResponseDto()
    sut = MainPresenter(view, repository, provider)
    `when`(repository.getTrendingResults(eq(DEFAULT_LIMIT_COUNT), eq(next)))
      .thenReturn(Observable.just(response))

    // Act
    sut.loadTrendingImages(next)

    // Assert
    verify(view).isActive()
    verify(view).addImages(eq(response))
  }

  @Test fun testLoadSearchImagesSuccess() {
    // Arrange
    val searchString = "gifs"
    val next = 0
    val response = RiffsyResponseDto()
    sut = MainPresenter(view, repository, provider)
    `when`(repository.getSearchResults(eq(searchString), eq(DEFAULT_LIMIT_COUNT), eq(next)))
      .thenReturn(Observable.just(response))

    // Act
    sut.loadSearchImages(searchString, next)

    // Assert
    verify(view).isActive()
    verify(view).addImages(eq(response))
  }

  @Test fun testLoadSearchImagesViewInactive() {
    // Arrange
    `when`(view.isActive()).thenReturn(false)
    sut = MainPresenter(view, repository, provider)
    val response = RiffsyResponseDto()
    val observable = Observable.just(response)

    // Act
    sut.loadImages(observable)

    // Assert
    verify(view, never()).addImages(any())
  }
}
